Mathlib/Data/Nat/Sqrt.lean

```diff
@@ -0,0 +1,319 @@
+  q * q ≤ n ∧ n < (q + 1) * (q + 1)
+-- #align nat.is_sqrt Nat.IsSqrt
+
+attribute [-simp] mul_eq_mul_left_iff mul_eq_mul_right_iff
+
+private theorem sqrtAux_IsSqrt_lemma (m r n : ℕ) (h₁ : r * r ≤ n) (m')
+    (hm : shiftr (2 ^ m * 2 ^ m) 2 = m')
+    (H1 : n < (r + 2 ^ m) * (r + 2 ^ m) → IsSqrt n (sqrtAux m' (r * 2 ^ m) (n - r * r)))
+    (H2 :
+      (r + 2 ^ m) * (r + 2 ^ m) ≤ n →
+        IsSqrt n (sqrtAux m' ((r + 2 ^ m) * 2 ^ m) (n - (r + 2 ^ m) * (r + 2 ^ m)))) :
+    IsSqrt n (sqrtAux (2 ^ m * 2 ^ m) (2 * r * 2 ^ m) (n - r * r)) := by
+  have b0 : 2 ^ m * 2 ^ m ≠ 0 := mul_self_ne_zero.2 (pow_ne_zero m two_ne_zero)
+  have lb : n - r * r < 2 * r * 2 ^ m + 2 ^ m * 2 ^ m ↔ n < (r + 2 ^ m) * (r + 2 ^ m) := by
+    rw [tsub_lt_iff_right h₁]
+    simp [left_distrib, right_distrib, two_mul, mul_comm, mul_assoc, add_comm, add_assoc,
+      add_left_comm]
+  have re : div2 (2 * r * 2 ^ m) = r * 2 ^ m := by
+    rw [div2_val, mul_assoc, Nat.mul_div_cancel_left _ (by decide : 2 > 0)]
+  cases' lt_or_ge n ((r + 2 ^ m) * (r + 2 ^ m)) with hl hl
+  · rw [sqrtAux_2 b0 (lb.2 hl), hm, re]
+    apply H1 hl
+  · cases' le.dest hl with n' e
+    rw [@sqrtAux_1 (2 * r * 2 ^ m) (n - r * r) (2 ^ m * 2 ^ m) b0 (n - (r + 2 ^ m) * (r + 2 ^ m)),
+      hm, re, ← right_distrib]
+    · apply H2 hl
+    apply Eq.symm
+    apply tsub_eq_of_eq_add_rev
+    rw [← add_assoc, (_ : r * r + _ = _)]
+    exact (add_tsub_cancel_of_le hl).symm
+    simp [left_distrib, right_distrib, two_mul, mul_comm, mul_assoc, add_assoc]
+-- #align nat.sqrt_aux_is_sqrt_lemma Nat.sqrtAux_IsSqrt_lemma
+
+private theorem sqrtAux_IsSqrt (n) :
+    ∀ m r,
+      r * r ≤ n →
+        n < (r + 2 ^ (m + 1)) * (r + 2 ^ (m + 1)) →
+          IsSqrt n (sqrtAux (2 ^ m * 2 ^ m) (2 * r * 2 ^ m) (n - r * r))
+  | 0, r, h₁, h₂ => by
+    apply sqrtAux_IsSqrt_lemma 0 r n h₁ 0 rfl <;> intro h <;> simp <;> [exact ⟨h₁, h⟩,
+      exact ⟨h, h₂⟩]
+  | m + 1, r, h₁, h₂ => by
+    apply
+        sqrtAux_IsSqrt_lemma (m + 1) r n h₁ (2 ^ m * 2 ^ m)
+          (by simp [shiftr, pow_succ, div2_val, mul_comm, mul_left_comm]) <;>
+      intro h
+    · have := sqrtAux_IsSqrt _ m r h₁ h
+      rw [mul_comm 2, mul_assoc, mul_comm 2] at this
+      exact this
+    · rw [pow_succ, mul_two, ← add_assoc] at h₂
+      have := sqrtAux_IsSqrt _ m (r + 2 ^ (m + 1)) h h₂
+      rw [mul_comm 2, mul_assoc, mul_comm 2] at this
+      exact this
+-- #align nat.sqrt_aux_is_sqrt nat.sqrtAux_IsSqrt
+
+-- Porting note: as the definition of square root has changed,
+-- the proof of `sqrt_IsSqrt` is attempted from scratch.
+/-
+Sketch of proof:
+Up to rounding, in terms of the definition of `sqrt.iter`,
+
+* By AM-GM inequality, `next² ≥ n` giving one of the bounds.
+* When we terminated, we have `guess ≥ next` from which we deduce the other bound `n ≥ next²`.
+
+To turn this into a lean proof we need to manipulate, use properties of natural number division etc.
+-/
+private theorem sqrt_IsSqrt (n : ℕ) : IsSqrt n (sqrt n) := by
```

Comment:
```suggestion
private theorem sqrt_isSqrt (n : ℕ) : IsSqrt n (sqrt n) := by
```

Reply:
Thanks. I wasn't sure about the naming conventions here, since the function is called `IsSqrt`.